contracts/middleware/BaseMiddleware.sol

```diff
@@ -0,0 +1,39 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.24;
+
+import {Hooks} from "@uniswap/v4-core/src/libraries/Hooks.sol";
+import {IPoolManager} from "@uniswap/v4-core/src/interfaces/IPoolManager.sol";
+import {IHooks} from "@uniswap/v4-core/src/interfaces/IHooks.sol";
+import {BalanceDelta} from "@uniswap/v4-core/src/types/BalanceDelta.sol";
+import {PoolKey} from "@uniswap/v4-core/src/types/PoolKey.sol";
+import {BeforeSwapDelta} from "@uniswap/v4-core/src/types/BeforeSwapDelta.sol";
+import {Proxy} from "@openzeppelin/contracts/proxy/Proxy.sol";
+
+contract BaseMiddleware is Proxy {
+    /// @notice The address of the pool manager
+    IPoolManager public immutable manager;
+    address public immutable implementation;
+
+    error FlagsMismatch();
+
+    constructor(IPoolManager _manager, address _impl) {
```

Comment:
what is the purpose of this BaseMiddleware? It seems to be mostly just a proxy wrapper. Are there any useful things that we can add for downstream implementers? Some ideas - 
- Standard modifiers i.e. onlyPoolManager at least?
- Standard implementations of some hook functions, with virtual functions that downstream implementers can override with their custom checks?

Reply:
it's pretty much just oz proxy yeah.
- i think it's up to the implementation to add the onlyPoolManager modifier. most hooks already use it, so adding onlyPoolManager to the middleware itself would result in a redundant check.
- like a beforeBeforeSwap, afterBeforeSwap pattern? this sounds a bit complicated and also creates deployment size bloat for middlewares that don't implement the hook.